src/contrib/mlir/core/ngraph_dialect/dialect.cpp

```diff
@@ -39,63 +39,60 @@ NGraphOpsDialect::NGraphOpsDialect(mlir::MLIRContext* ctx)
         >();
 }
 
-mlir::Type NGraphOpsDialect::parseType(llvm::StringRef tyData, mlir::Location loc) const
+mlir::Type NGraphOpsDialect::parseType(mlir::DialectAsmParser& parser) const
 {
-    StringRef origTypeStr = tyData;
     MLIRContext* context = getContext();
 
     // Process nGraph tensor type.
-    if (tyData.consume_front("tensor"))
+    // failure is true
+    if (!parser.parseOptionalKeyword("tensor"))
     {
-        if (!tyData.consume_front("<") || !tyData.consume_back(">"))
+        llvm::SMLoc typeLoc = parser.getCurrentLocation();
+        if (parser.parseLess())
         {
-            return (emitError(loc, "expected '<' and '>' enclosing the tensor shape: " + tyData),
-                    Type());
+            parser.emitError(typeLoc, "expected '<' and '>' enclosing the tensor shape");
+            return Type();
         }
 
-        // Get x-separated sub-strings.
-        SmallVector<StringRef, 8> subStrings;
-        tyData.split(subStrings, "x");
-
         // Parse shape dimensions.
         SmallVector<int64_t, 4> shape;
-        for (unsigned i = 0, end = subStrings.size() - 1; i < end; ++i)
-        {
-            StringRef dimStr = subStrings[i];
-            int64_t dim = -1;
-            // NOTE: `consumeInteger` returns false if an integer was parsed successfully.
-            if (dimStr.consumeInteger(/*Radix=*/10, dim) || !dimStr.empty())
-            {
-                return (
-                    emitError(loc, "expected a list of '[0-9]+x' dimension specifiers: " + tyData),
-                    Type());
-            }
-
-            shape.push_back(dim);
-        }
+        parser.parseDimensionList(shape);
 
-        // Parse nGraph element type.
-        auto elem_ty = mlir::parseType(subStrings.back(), context);
-        if (!elem_ty)
-        {
-            return (emitError(loc, "Unexpected element type in tensor type: " + tyData), Type());
-        }
+        // Parse the current element type.
+        Type eltType;
+
+        parser.parseType(eltType);
 
-        return NGTensorType::get(context, elem_ty, shape);
+        parser.parseGreater();
```

Comment:
Nice! These are great improvements! Should we keep error messages in lines 81 and 70?

Reply:
Right. Fixed.